src/libaktualizr/primary/sotauptaneclient.cc

```diff
@@ -546,12 +546,23 @@ result::Download SotaUptaneClient::downloadImages(const std::vector<Uptane::Targ
   if (!targets.empty()) {
     if (targets.size() == downloaded_targets.size()) {
       result = result::Download(downloaded_targets, result::DownloadStatus::kSuccess, "");
-    } else if (downloaded_targets.size() == 0) {
-      LOG_ERROR << "None of " << targets.size() << " targets were successfully downloaded.";
-      result = result::Download(downloaded_targets, result::DownloadStatus::kError, "Each target download has failed");
     } else {
-      LOG_ERROR << "Only " << downloaded_targets.size() << " of " << targets.size() << " were successfully downloaded.";
-      result = result::Download(downloaded_targets, result::DownloadStatus::kPartialSuccess, "");
+      if (downloaded_targets.size() == 0) {
+        LOG_ERROR << "None of " << targets.size() << " targets were successfully downloaded.";
+        result =
+            result::Download(downloaded_targets, result::DownloadStatus::kError, "Each target download has failed");
+      } else {
+        LOG_ERROR << "Only " << downloaded_targets.size() << " of " << targets.size()
+                  << " were successfully downloaded.";
+        result = result::Download(downloaded_targets, result::DownloadStatus::kPartialSuccess, "");
+      }
+      // Store installation report to inform Director of the download failure.
+      const std::string &correlation_id = director_repo.getCorrelationId();
+      data::InstallationResult device_installation_result =
+          data::InstallationResult(data::ResultCode::Numeric::kDownloadFailed, "Target download failed");
```

Comment:
As far as I understand the mechanism of cancelling/removing of the new targets at Director is a sending manifest to it with a negative installation result in case if all targets downloads fails ?

Reply:
If I'm understanding you correctly, yes. If we send an installation report with a failure code, the Director will respond by identifying the installation as a failure, and subsequent Targets metadata will be empty (to indicate that the device should take no further action, i.e. not retry until explicitly told to do so). Let me know if that doesn't make sense or you have further questions.